### PR TITLE
Clean/block validation

### PR DESCRIPTION
**Pull request by ebuchman**
Clean up the `MakeBlock` and `ValidateBasic` functions by distinguishing what can be done with just the Block, and what requires the State.

Also make some structure in the State more explicit.

### CODE REVIEW

File: state/execution_test.go [hunk 1]
```diff
@@ -23,6 +23,43 @@ var (
 	nTxsPerBlock = 10
 )
 
+/*
```
Comment (ebuchman) [inline state/execution_test.go]: TODO: test

File: state/state.go [hunk 2]
```diff
@@ -27,40 +28,51 @@ func calcValidatorsKey(height int64) []byte {
 	return []byte(cmn.Fmt("validatorsKey:%v", height))
 }
 
+func calcConsensusParamsKey(height int64) []byte {
```
Comment (ebuchman) [inline state/state.go]: TODO: actually use this to store the historical consensus params.

Alternatively, just stick the ConsensusParams directly in the header. But that's probably ridiculous.

File: types/block.go [hunk 3]
```diff
@@ -15,69 +15,38 @@ import (
...
-// TODO: Add version information to the Block struct.
-func MakeBlock(height int64, chainID string, txs []Tx,
-	totalTxs int64, commit *Commit,
-	prevBlockID BlockID, valHash, appHash, consensusHash []byte,
-	partSize int) (*Block, *PartSet) {
-
-	newTxs := int64(len(txs))
+// MakeBlock returns a new block with an empty header, except what can be computed from itself.
+// It populates the same set of fields validated by ValidateBasic
+func MakeBlock(height int64, txs []Tx, commit *Commit) *Block {
```
Comment (ebuchman) [inline types/block.go]: maybe send chainID in here too ?

File: types/block.go [hunk 4]
```diff
@@ -15,69 +15,38 @@ import (
...
 		Data: &Data{
 			Txs: txs,
 		},
 	}
 	block.FillHeader()
-	return block, block.MakePartSet(partSize)
+	return block
 }
 
 // ValidateBasic performs basic validation that doesn't involve state data.
```
Comment (ebuchman) [inline types/block.go]: lol at this comment prior to this PR :P

File: state/state.go [hunk 5]
```diff
@@ -27,40 +28,51 @@ func calcValidatorsKey(height int64) []byte {
...
+	Validators                  *types.ValidatorSet
+	LastValidators              *types.ValidatorSet
 	LastHeightValidatorsChanged int64
 
-	// AppHash is updated after Commit
+	// Consensus parameters used for validating blocks.
+	// Changes returned by EndBlock and updated after Commit.
+	ConsensusParams                  types.ConsensusParams
+	LastConsensusParams              types.ConsensusParams
+	LastHeightConsensusParamsChanged int64
```
Comment (melekes) [inline state/state.go]: don't see where we actually change it and use it.

Comment (ebuchman) [reply state/state.go]: still need to save these and use this ...

**Pull request merged**